### ISSUE TITLE
Deceleration When Not Moving

### ISSUE THREAD

**Issue by account3r2** [RobertCochran/momentum-engine#1]
# Description:  
When against a wall and continuing to move in the direction that the wall is, you still seem to have deceleration when you stop holding the move key.  
  
# Expected Result:  
Moving against a wall does nothing.  
  
# Actual Result:  
Moving against a wall will make you fly backwards after you release the move key.

**Comment by RobertCochran** [comment]
Yeah... It's a result of a sync issue between client and server. When you hit
up against a wall, the server resets the X/Y movement variable (see
Player.update() in src/server/player.py). The client remains unaware of this,
so when the key release is sent back to the player, it causes movement in the
opposite direction (Moving in X direction -1, hit wall, reset X movement to 0
, key release, set X movement to 1, causing you to go eastward). Unfortunately,
I'm not aware of a way at the moment to fix this without breaking the rest of
the movement logic. Perhaps its necessary to keep some local state in the
client... I'm actually going to test with two separate machines running the two
halves, rather than one machine running both as I have up to this point. During
this testing, I'm going to determine for myself if having shared logic state is
necessary. If so, it will be fixed when that is implemented. Otherwise, I'm
unsure of how to fix this.

**Comment by JoshuaBrockschmidt** [comment]
I feel it'd be a bit clunky to have the client handle any of the collision detection. As well, it could present some serious security flaws. Maybe just have the server tell the client how much they actually moved, (or even just whether they moved at all)?

**Comment by JoshuaBrockschmidt** [comment]
I might toy with it myself if I can find the time to do so.

**Comment by RobertCochran** [comment]
The problem in having the server trying to do *everything* is the sheer amount
of network traffic that occurs. It only works on a local system because there
is no actual packet travel and the communication effectively becomes shared
memory, which improves performance *significantly*. I'm not kidding, there is
at /least/ 120 packets/sec between a single client and the server, because
currently every client cycle (limited to 1/60 s) sends a packet out to the
server, which then responds. And I'm talking about just the player update
retrieval packet. This doesn't factor in the packets that get sent out to
notify the server of keypress changes or anything like that. There's just too
much network traffic with the current model. And as far as I can tell (and I
may just be missing something very obvious), the only way to fix it is to send
packets out less often. And the only way that I can see to fix *that* is to have
the client keep track of things locally and send a delta to the server, say
every 1/2 second or so. If you think I'm missing something that you feel is
obvious, or you have an epiphany, by all means tell me. But from what
(admittedly little) testing I did, performance slowed significantly when client
and server did not reside on the same server, and I'm willing to bet that
the amount of traffic being generated was the culprit.

tl;dr - network model is currently broken, and the only way I see to fix is to
have the client keep track of some data so less packets can be sent out,
instead of relying on server for everything

**Comment by account3r2** [comment]
I see. I think that around 20 packets/sec per client would be ideal, but I can't be of much help without going a bit in-depth with your project, which I don't have enough free time to take a look at for now.

**Comment by RobertCochran** [comment]
With some insight from the future, performance testing has indicated that
packet transfer rate is actually a non-issue. With that out of the way, testing
has /also/ indicated that CPU, memory, and bandwidth usage are also more than
fine, even considering the complexity (or lack thereof) of the engine currently.

Now, code-wise, the obvious solution is to have the client check when receiving
a packet from the server whether or not the xmovement or ymovement variable has
changed remotely (because the client gets the entirety of the class's objects
every time it requests an update). This is a fairly straightforward change.

**Comment by RobertCochran** [comment]
Fixed in commit b0cdb45dbb226af60dcfce0aece86a57ee939048
